pcl_detector/src/pcl_detector_ros.cpp

```diff
@@ -0,0 +1,143 @@
+  qos_profile.reliability = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
+  auto qos = rclcpp::QoS(rclcpp::QoSInitialization(qos_profile.history, 1), qos_profile);
+  
+  publisher_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(param_topic_pointcloud_out_, qos);
+  subscription_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
+    param_topic_pointcloud_in_, qos, std::bind(&PclDetectorNode::topic_callback, this, _1));
+
+  // Define a handle for validating parameters during runtime  
+  on_set_callback_handle_ = add_on_set_parameters_callback(std::bind(&PclDetectorNode::parametersCallback, this, std::placeholders::_1));
+  
+  // Initialize the detector with configured parameters
+  std::string detector;
+  get_parameter<std::string>("detector", detector);
+  m_detector = initialize_detector(detector);
+}
+
+// Callback function for parameter changes
+rcl_interfaces::msg::SetParametersResult PclDetectorNode::parametersCallback(const std::vector<rclcpp::Parameter> &parameters) {
+    std::string detector_name;
+    rcl_interfaces::msg::SetParametersResult result;
+
+    for (const auto &parameter : parameters) {
+        if (parameter.get_name() == "detector") {
+            if (detector_type.count(parameter.as_string()) == 0) {
+                result.successful = false;
+                return result;
+            }
+        }
+    }
+    
+    parameters_changed = true;
+    result.successful = true;
+    result.reason = "success";
+    return result;
+}
+
+// Initialize the selected detector based on the parameter
+std::unique_ptr<IPclDetector> PclDetectorNode::initialize_detector(std::string detector)
+{
+    std::unique_ptr<IPclDetector> configured_detector = nullptr;
+
+    switch (detector_type[detector]) {
+        case DetectorType::Euclidean: {
+            float cluster_tolerance;
+            get_parameter<float>("euclidean.cluster_tolerance", cluster_tolerance);
+            int min_points;
+            get_parameter<int>("euclidean.min_points", min_points);
+            configured_detector = std::make_unique<EuclideanClusteringDetector>(cluster_tolerance, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "Euclidean detector initialized with cluster tolerance = " << cluster_tolerance << " and min_points = " << min_points);
+            break;
+        }
+
+        case DetectorType::DBSCAN: {
+            float eps; 
+            get_parameter<float>("dbscan.epsilon", eps);
+            int min_points;
+            get_parameter<int>("dbscan.min_points", min_points);
+            configured_detector = std::make_unique<DBSCANDetector>(eps, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "DBSCAN detector initialized with eps = " << eps << " and min_points = " << min_points);
+            break;
+        }
+
+        default: {
+            RCLCPP_ERROR(this->get_logger(), "Invalid pcl detector specified! Shutting down...");
+            rclcpp::shutdown();
```

Comment:
Is it necessary to shut down? Is it possible to throw a warning and reject the new value and revert to the previous one instead? If this happens on initialization and we don't have a default then shut down of course

Reply:
This is handled by the reconfigure_callback. The node will only be shut down if the initial parameters are invalid :)